### PR TITLE
Add the API implementation for the relationship widget (#33).

### PR DESCRIPTION
**Pull request by chaoran-chen**
Addressing #33, I've implemented a part of the API for the relationship widget.

@lustoykov, @martomi: Please take a look and tell me if it fits your requirements. I've chosen some other names than those in #33 because it seems to me that names like `subject` are more common used through our project than `primaryEntity`. But of course, I can change it. Just tell me every wish you have :)

@MusicConnectionMachine/group-3, @MusicConnectionMachine/group-4: I have no idea how to map an artist / work to a subject. Can you take care of it? ([Cf., code.](https://github.com/MusicConnectionMachine/api/compare/feature/api-for-relationship-widget?expand=1#diff-f3324d286d2ffee1c4b726eb65bdc3dcR8))

And what's about the autosuggestion? How should it be implemented? ([Cf., code.](https://github.com/MusicConnectionMachine/api/compare/feature/api-for-relationship-widget?expand=1#diff-f3324d286d2ffee1c4b726eb65bdc3dcR47))

### CODE REVIEW

File: api/dsap/relationships.js [hunk 1]
```diff
@@ -1,22 +1,67 @@
...
+                    { model: models.relationshipOccurrences },
+                    { model: models.relationshipEntities, as: 'Subject', where: subjectWhere},
+                    { model: models.relationshipEntities, as: 'Object', where: objectWhere }
+                ],
+                offset,
+                limit
+            });
+        },
+        findSuggestions: function (subject, query, offset = 0, limit = 20) {
+            // TODO Perform an actual search (Using Elasticsearch? PostgreSQL?)
```
Comment (pfent) [inline api/dsap/relationships.js:57]: This should probably be a postgres full-text search

Comment (Henni) [reply api/dsap/relationships.js:57]: @pfent :+1: , do you have any experience with this (and maybe even on how to do this in sequelize in a not completely ugly way)? I already experimented with some `ts_query` stuff and found a few helpful pages, but wasn't able to build anything useful with the sequelize syntax.

File: api/controllers/relationships.js [hunk 2]
```diff
@@ -20,6 +24,51 @@ function getAllRelationships(req, res) {
...
+}
+
+function _getAllRelationShipsBy(req, res, mappingFunction) {
+    var params = req.swagger.params;
+    var id = params.id.value;
+    var offset = params.offset ? params.offset.value : undefined;
+    var limit = params.limit ? params.limit.value : undefined;
+    var subject = mappingFunction(id);
+    var relation = params.relation ? params.relation.value : undefined;
+    var object = params.object ? params.object.value : undefined;
```
Comment (Henni) [inline api/controllers/relationships.js]: You can directly use `params.object.value` which evaluates to `undefined` if not given as parameter.

**Review by Henni — CHANGES_REQUESTED** [review]

File: api/dsap/relationships.js [hunk 3]
```diff
@@ -1,22 +1,67 @@
...
+    var models = context.models;
 
-    return {
+    const functions = {
+        mapArtistToSubject: function (aritistId) {
+            // TODO Map artist to subject.
+            return 'A0EEBC99-9C0B-4EF8-BB6D-6BB9BD380A11';
+        },
+        mapWorkToSubject: function (workId) {
+            // TODO Map work to subject.
```
Comment (Henni) [inline api/dsap/relationships.js]: I'm looking into this.

File: api/controllers/relationships.js (hunk 2 above)
Comment (chaoran-chen) [reply api/controllers/relationships.js]: Ah, cool. Thanks!

File: api/dsap/relationships.js (hunk 1 above)
Comment (pfent) [reply api/dsap/relationships.js:57]: @Henni Maybe for starters, you can use [pg_trgm](https://www.postgresql.org/docs/current/static/pgtrgm.html) to get suggestions for words. Just `sequelize.fn` the `similarity`, and get the top 5…

I'm not really sure, how well `ts_query` can map to sequelize…

File: api/dsap/relationships.js (hunk 3 above)
Comment (Henni) [reply api/dsap/relationships.js]: Something like this should work:
```javascript
models.artists.findById(artistId)
  .then(artist => {
    return artist.getEntity();
  })
  .then(entity => {
    return entity.getRelEntity();
  });
```
This is pretty ugly as it creates three queries.

There should also be another way, like this:
```javascript
relEntities.findOne({
                attributes: ['id'],
                include: [{
                    model: entities,
                    include: [{
                        model: models.artists
                    }]
                }],
                where: {
                    '$artists.id$': artistId
                }
            })
```

This already includes some sequelize magic with the `$...$` which allows one to build a where clause on an included table, which is documented nowhere except for some Github issue. But it doesn't work and the query grows pretty large.

Maybe some Sequelize magician can jump in and fix this abomination of query...

Comment (chaoran-chen) [reply api/dsap/relationships.js]: You can just commit it if it works :)

**Review by Henni — APPROVED** [review]

**Comment by Henni** [comment]
I added the mapper functions which *should* work. I wasn't really able to test it, as my local database currently is somehow broken...

Please still test this before merging!

**Comment by chaoran-chen** [comment]
Thanks @Henni! Can someone, who actual has some real data, test it, please?

**Comment by pfent** [comment]
@MusicConnectionMachine/group-3 @MusicConnectionMachine/group-4 do you have some data to test this?

**Pull request merged**